core/sql/optimizer/SynthType.cpp

```diff
@@ -5207,6 +5207,18 @@ const NAType *Translate::synthesizeType()
          err4106arg = SQLCHARSETSTRING_UTF8;
        break;
 
+     case GBK_TO_UTF8:
+       if (translateSource->getCharSet() == CharInfo::GBK || translateSource->getCharSet() == CharInfo::UnknownCharSet )
+         charsetTarget = CharInfo::UTF8;
+       else
+       {
+            if( CmpCommon::getDefaultString(HIVE_FILE_CHARSET) ==  CmpCommon::getDefaultString(HIVE_DEFAULT_CHARSET) )
+              err4106arg = SQLCHARSETCODE_GB2312;
+            else
+             charsetTarget = CharInfo::UTF8;
+       }
+       break;
+
```

Comment:
Note that we also need to calculate the length of the new type, that is done below by calling CharInfo::getMaxConvertedLenInBytes(). Again, the optimization I mentioned on line 9327 would be useful here, to avoid wasting space. However, I see a problem in Translate::synthType here, in that it always generates an output type with a size specified in characters, not bytes. That's inefficient and we should consider fixing it. For GBK->UTF8, this means that a 1000 byte input buffer will result in a 4000 byte output buffer, even though we can show that 1500 bytes would be enough.